docs/OBJECT_LOCK_TEST_PLAN.md

```diff
@@ -0,0 +1,157 @@
+ObjectLockConfigurationNotFoundError error
+- getting configuration returns correct object lock configuration for bucket
+
+### putObjectRetention
+
+- putting retention as non-bucket-owner user returns AccessDenied error
+- putting retention on object in bucket without object lock enabled returns
+InvalidRequest error
+- putting valid retention period updates object metadata
+
+### getObjectRetention
+
+- getting retention as non-bucket-owner user returns AccessDenied error
+- getting retention when none is set returns NoSuchObjectLockConfiguration
+error
+- getting retention returns correct object retention period
+
+### putObjectLegalHold
+
+- putting legal hold as non-bucket-owner user returns AccessDenied error
+- putting legal hold on object in bucket without object lock enabled returns
+InvalidRequest error
+- putting valid legal hold updates object metadata
+
+### getObjectLegalHold
+
+- getting legal hold as non-bucket-owner user returns AccessDenied error
+- getting legal hold when none is set returns NoSuchObjectLockConfiguration
+error
+- getting legal hold returns correct object legal hold
+
+## End to End Tests
+
+### Scenarios
+
+- Create bucket with object lock enabled. Put object. Put object lock
+configuration. Put another object.
+
+    - Ensure object put before configuration does not have retention period set
+    - Ensure object put after configuration does have retention period set
+
+- Create bucket without object lock. Put object. Enable object lock with token
+and put object lock configuration. Put another object.
+
+    - Ensure object put before configuration does not have retention period set
+    - Ensure object put after configuration does have retention period set
+
+- Create bucket with object lock enabled and put configuration with COMPLIANCE
+mode. Put object.
+
+    - Ensure object cannot be deleted (returns AccessDenied error).
+    - Ensure object cannot be overwritten.
+
+- Create bucket with object lock enabled and put configuration with GOVERNANCE
+mode. Put object.
+
+    - Ensure user without permission cannot delete object
+    - Ensure user without permission cannot overwrite object
+    - Ensure user with permission can delete object
+    - Ensure user with permission can overwrite object
+    - Ensure user with permission can lengthen retention period
+    - Ensure user with permission cannot shorten retention period
+
+- Create bucket with object lock enabled and put configuration with date past
+expiration. Put object.
+
+    - Ensure object can be deleted.
+    - Ensure object can be overwritten.
+
+- Create bucket with object lock enabled and put configuration with date past
+expiration. Put object. Put new retention period on object.
+
+    - Ensure object cannot be deleted.
+    - Ensure object cannot be overwritten.
+
+- Create bucket with object locked enabled and put configuration. Put object.
+Put new retention period with date past expiration on object.
```

Comment:
I think this test could be misleading as well, as putting object with retention period with date past expiration returns `An error occurred (InvalidArgument) when calling the PutObject operation: The retain until date must be in the future!` error by AWS.

Reply:
I haven't tested it fully to know exactly whether that error is returned by the sdk itself, but we should still be testing scenarios with dates past expiration to make sure it doesn't produce any unexpected behaviors :)

Reply:
Yeah that's right; we'd still wanna test it to verify the behavior in that scenario :)